Dart/src/com/jetbrains/lang/dart/analyzer/DartAnalysisServerService.java

```diff
@@ -742,6 +739,80 @@ public void run() {
     return resultRef.get();
   }
 
+  public void search_findElementReferences(@NotNull final String filePath,
+                                           final int offset,
+                                           @NotNull final Processor<SearchResult> processor) {
+    final String searchId;
+    synchronized (myLock) {
+      if (myServer == null) return;
+      final AnalysisServer server = myServer;
+
+      final Ref<String> searchIdRef = new Ref<String>();
+      final Semaphore semaphore = new Semaphore();
+
+      semaphore.down();
+      final boolean ok = runInPooledThreadAndWait(new Runnable() {
+        @Override
+        public void run() {
+          server.search_findElementReferences(filePath, offset, true, new FindElementReferencesConsumer() {
+            @Override
+            public void computedElementReferences(String searchId, Element element) {
+              searchIdRef.set(searchId);
+              semaphore.up();
+            }
+
+            @Override
+            public void onError(RequestError requestError) {
+              semaphore.up();
+            }
+          });
+        }
+      }, "search_findElementReferences(" + filePath + ", " + offset + ")", SEND_REQUEST_TIMEOUT);
+
+      if (!ok) {
+        stopServer();
+        return;
+      }
+
+      final long t0 = System.currentTimeMillis();
+      semaphore.waitFor(FIND_ELEMENT_REFERENCES_TIMEOUT);
+
+      if (semaphore.tryUp()) {
+        LOG.info("search_findElementReferences() took too long for file " +
+                 filePath +
+                 "@" +
+                 offset +
+                 ": " +
+                 (System.currentTimeMillis() - t0) +
+                 "ms");
+        return;
+      }
+
+      searchId = searchIdRef.get();
+    }
+
+    while (true) {
+      synchronized (mySearchResultSets) {
+        SearchResultsSet resultSet;
+        // process already received results
+        while ((resultSet = mySearchResultSets.poll()) != null) {
+          if (!resultSet.id.equals(searchId)) continue;
+          for (final SearchResult searchResult : resultSet.results) {
+            processor.process(searchResult);
+          }
+          if (resultSet.isLast) return;
+        }
+        // wait for more results
+        try {
+          mySearchResultSets.wait();
```

Comment:
ProgressManager.checkCancelled() is extremely important for user experience (instant reaction on Escape/Cancel button). See how it is done in addCompletions()